### ISSUE TITLE
Does not work in Chromium (Snap).

### ISSUE THREAD

**Issue by joshcangit** [keepassxreboot/keepassxc-browser#405]
## Current Behavior

KeePassXC-Browser in Chromium won't connect to KeePassXC even when database is open and Chromium is checked.
![keepassxc-chromium](https://user-images.githubusercontent.com/33022160/52166293-4951e600-2703-11e9-812e-48692fd62067.png)
![keepassxc-integration](https://user-images.githubusercontent.com/33022160/52166295-4d7e0380-2703-11e9-9bbe-c0a9554ca245.png)

## Expected Behavior

KeePassXC-Browser works in Firefox.
![keepassxc-firefox](https://user-images.githubusercontent.com/33022160/52166323-bd8c8980-2703-11e9-8b46-ce61e634aa07.png)

## Steps to Reproduce (for bugs)

1. Install KeePassXC on Linux and use or create a database.
2. Enable Browser Integration for Chromium.
3. Install Chromium on Linux from snapcraft.io.
4. Install KeePassXC-Browser extension from the Chrome Web Store.
5. Under Connected Databases in the settings of KeePassXC-Browser, click `Connect`.

## Debug info

KeePassXC:
```
KeePassXC - Community-driven port of the Windows application “KeePass Password
Safe”

          ID: org.keepassxc.KeePassXC
         Ref: app/org.keepassxc.KeePassXC/x86_64/stable
        Arch: x86_64
      Branch: stable
     Version: 2.3.4
     License: GPL-3.0+
      Origin: flathub
  Collection: org.flathub.Stable
Installation: system
   Installed: 13.5 MB
     Runtime: org.kde.Platform/x86_64/5.12
         Sdk: org.kde.Sdk/x86_64/5.12

      Commit: 03a58a317bdf7107cde8f31c30bc5bb0fd4dd2e5d8761b7df324e84c5bda24c9
      Parent: 8702d16b321c4ed61f14f7e9dbbccea84cef01f369b474655f680277ddc79bb6
     Subject: Update sdk to 5.12 (0ba59621)
        Date: 2019-01-22 15:52:59 +0000
```
KeePassXC-Browser Version: _1.3.2_
Operating system: Linux
```
Description:	Ubuntu 18.04.1 LTS
```
Snap:
```
snap    2.37.1
snapd   2.37.1
series  16
ubuntu  18.04
```
Flatpak:
```
Flatpak 1.2.0
```
Browser: Chromium
```
Name      Version       Rev  Tracking  Publisher   Notes
chromium  71.0.3578.98  566  stable    canonical*  -
```
Browser: Firefox
```
Mozilla Firefox 65.0
```

**Comment by droidmonkey** [comment]
Using a Snap version of a browser requires you to manually embed the configuration file. This is because snaps install configuration files in non-standard locations.

**Comment by joshcangit** [comment]
Well, how would it works?

**Comment by droidmonkey** [comment]
You need to move ```~/.config/chromium/NativeMessagingHosts/org.keepassxc.keepassxc_browser.json``` to ```~/snap/chromium/current/.config/chromium/NativeMessagingHosts/```

You will have to create the NativeMessagingHosts folder.

**Comment by joshcangit** [comment]
Actually, that didn't solved it.

**Comment by droidmonkey** [comment]
I recommend using autotype coupled with an extension that changes the window title based on the website you are on.

**Comment by twt2** [comment]
Just adding a comment that as of Ubuntu 19.10 eoan Chromium is only available as a snap.

**Comment by joshcangit** [comment]
That's disappointing. Chromium from 19.10 and onwards will need to use standalone add-ons such as [Keepass Tusk](https://subdavis.com/Tusk/) to work but worse is that add-on is no longer maintained so the only solutions are to build more of such add-ons or just rely on using Auto-Type.

**Comment by HerKa2** [comment]
The situation (lake of solution to have a password safe working with Chromium on Ubuntu ) means that I will not use Chromium any more.

**Comment by droidmonkey** [comment]
For context, this isn't just impacting us. Any extension that uses native messaging *will not work* in a browser installed as a snap. The snap developers need to add a plug for native messaging or similar.

**Comment by FMuro** [comment]
Would it be possible to make this work if Chromium is installed with classic confinement? (Which in practice means unconfined.) I'd be glad to test it if I knew what I should do to check it.

**Comment by droidmonkey** [comment]
Possibly yes

**Comment by FMuro** [comment]
> Possibly yes

I would be glad if you could give me a hint on how to try it.

**Comment by FMuro** [comment]
Sorry to say I asked the snapcrafters and it doesn't seem to be possible to install Chromium with classic confinement, so we'll have to wait until they fix whatever is needed in the snaps.

**Comment by derflocki** [comment]
As others have pointed out, there doesn't seem to be a fix given the state of snaps.

I have found a workaround however:

## chromium-snap "without" the snap
This hack directly runs chromium from the snap. So I it might not be as secure as running it within the snap.

* Overrule `chromium_chromium.desktop` from snap:


    * Copy snap desktop file to ~/.local/share/applications
    * `cp /var/lib/snapd/desktop/applications/chromium_chromium.desktop ~/.local/share/applications/chromium_chromium.desktop`
    * apply [snap-to-local.diff.txt](https://github.com/keepassxreboot/keepassxc-browser/files/5338740/snap-to-local.diff.txt), basically only the Exec statements need to change, but to make it work nicely also **Icon**, **X-SnapInstanceName** and **StartupWMClass**
    * or use [chromium_chromium.desktop.txt](https://github.com/keepassxreboot/keepassxc-browser/files/5338743/chromium_chromium.desktop.txt)

* create symlink to `/usr/local/bin/chromium`
** `sudo ln -sf /snap/chromium/current/usr/lib/chromium-browser/chrome /usr/local/bin/chromium`

I have been running this for a couple of weeks now, on 2 machines and it works just fine. I have encountered 2 minor issues:

1. on the one machine, chromium kept asking to be the default browser again
2. on the other machine, I have the pinned icon in the launcher and another  icon for the currentl runing instance. But maybe I have just made a typo in the **StartupWMClass**.

I hope this helps.

**Comment by FMuro** [comment]
@derflocki oh, man! That works! Just wanted to confirm.

**Comment by cptechnik** [comment]
save 
> or use [chromium_chromium.desktop.txt](https://github.com/keepassxreboot/keepassxc-browser/files/5338743/chromium_chromium.desktop.txt)
on
~/.local/share/applications/chromium_chromium.desktop
and start chromium with double-click on this.

fine.

**Comment by alexswerner** [comment]
Hi,

here a script to fix this issue in a couple of steps. Same what @derflocki and @cptechnik posted, just adds your user name right in the right place. 

```
sudo ln -sf /snap/chromium/current/usr/lib/chromium-browser/chrome /usr/local/bin/chromium

cp /var/lib/snapd/desktop/applications/chromium_chromium.desktop ~/.local/share/applications/chromium_chromium.desktop
sed -i "s/\/var\/lib\/snapd\/desktop\/applications\/chromium_chromium.desktop/\/home\/$(whoami)\/.local\/share\/applications\/chromium_chromium.desktop/" ~/.local/share/applications/chromium_chromium.desktop
sed -i "s/\/snap\/bin\/chromium/\/snap\/chromium\/current\/usr\/lib\/chromium-browser\/chrome/" ~/.local/share/applications/chromium_chromium.desktop
sed -i "s/Icon=\/snap\/chromium\/.*\/chromium.png/Icon=\/snap\/chromium\/current\/chromium.png/"  ~/.local/share/applications/chromium_chromium.desktop
sed -i "s/StartupWMClass=chromium/StartupWMClass=chromium-browser/"  ~/.local/share/applications/chromium_chromium.desktop
```

**Comment by droidmonkey** [comment]
That completely invalidates the snap. Don't offer this advice, it's not accurate.

**Comment by alexswerner** [comment]
@droidmonkey I understand that this is what happens, but what is the better solution here? I'm not super familiar with snap, so please bear with me. My options are:
* Install chromium (not google-chrome) from a ppa which comes with the requirement to also add an ffmpeg ppa which breaks so many dependencies.
* This. From my understanding snap is providing some kernel-based confinement so that even if the web browser is sandbox is compromised, the attacker cannot access the users files. Or in an alternative scenario where the application is the attacker, it cannot access the users files. For the first case, I will just hope that the browser sandbox is good enough.

Is there something I am missing here? For me chromium based snap is full of broken use cases (e.g. this and not being able to install gnome extensions any more). And adding an alternative package management just seems very wrong.

Any enlightenment is appreciated!

**Comment by droidmonkey** [comment]
You can install the snap with classic confinement https://snapcraft.io/docs/snap-confinement

**Comment by alexswerner** [comment]
Installing with classic confinement: That sounded sensible, but I also learned that this was not properly implemented. Also tried `snap install --devmode chromium` which does not work either.

Still not sure if _sudoing symlinks around_ is really worse than breaking the user experience in multiple ways.

**Comment by derflocki** [comment]
Personally I have switched to Firefox (because of this and other reasons). 

What I have posted is obviously a hack and might break without notice and - as i mentioned in my original message - could be a security risk.

IMHO this issue should be closed as afaik there is nothing keepassxc can do to make it work. Would be good to link the relevant chrome snap issue to further track the progress. Also it might be interesting if this works with flatpak (when i came up with the hack, they had the same issue).
